README.md
# combinator-FQ
Combinator-FQ is a tool for detecting adjacent contigs and calculating the LQ-coefficient. The LQ-coefficient reflects the degree of connection between contigs. The higher the LQ-coefficient, the higher is the degree of connection.

Latest version: `1.6.a`.

### Description

Combinator-FQ identifies adjacent contigs in order to facilitate further scaffolding.

Format of input: multi-fasta file containing contigs.

Combinator-FQ supports contigs assembled by SPAdes and A5.

It calculates LQ-coefficient, defined as follows:

![](formulas/LQ_definition.png)

Length of overlap is further referred to as *k*. *k-start* (*k-end*) denotes contig's prefix (suffix) of length *k*.

Ends of contigs are considered adjacent in following situations:

![](formulas/adjacent_contigs.png)

combinator-FQ calculates expected length of the genome as follows:

![](formulas/exp_genome_len.png)

If no coverage information is present in sequence headers, multiplicity is calculated based on discovered overlaps of the i-th contig as follows:

![](formulas/ovl_multplty.png)

### Dependencies

Combinator-FQ is written in Python, so you need Python interpreter (version 3.X) to use it. [Here you can download Python](https://www.python.org/downloads/).

### Explanation of output files

Combinator-FQ generates 3 output files:

#### 1) Table containing information about contigs adjacency:

    <prefix>_combinator_adjacent_contigs.tsv

Format of table is following:

```
#  Contig name  Length  Coverage  GC(%)  Multiplicity  Annotation  Start         End
1  NODE_1...    304356  93.7155   34.12  1.2           [empty]     [S=E(NODE_22); ovl=127]  [E=rc_E(NODE_26); ovl=99]
...
```
Explanation of "Start" and "End" columns:

- `[S=E(NODE_22); ovl=127]` means that k-start (`S`) of contig `NODE_1...` is identical to k-end (`E`) of contig `NODE_22`, and overlap length is 127 (`ovl=127`).

- `[E=rc_E(NODE_26); ovl=99]` means that k-end of contig `NODE_1...` is identical to reverse complement (`rc_`) k-end of contig `NODE_26`, and overlap length is 99 (`ovl=99`).

#### 2) File, in which all matchings (not only adjacency-associated) are listed:

    <prefix>_combinator_full_matching_log.txt

#### 3) Brief summary:

    <prefix>_combinator_summary.txt

### Options

```
-h (--help): print help message;

-v (--version): print version;

-i (--mink): minimum k (in b.p.) to consider.
  Value: integer > 0; Default if 21 b.p.

-a (--maxk): maximum k (in b.p.) to consider.
  Value: integer > 0; Default is 127 b.p.

-k (--k-mer-len): exact k (in b.p.).
  If specified, '-i' and '-a' options are ignored.
  Value: integer > 0; Option is disabled by default;

-o (--outdir): output directory;
  Default value: 'combinator-result'
```

### Examples

```
  ./combinator-FQ.py contigs.fasta -k 127 -o my_outdir
```

```
  ./combinator-FQ.py another_contigs.fa -i 25 -a 300 -o my_outdir
```

If input file is omitted in the command (like in the command below), combinator-FQ will process all fasta files in the working directory.

```
  ./combinator-FQ.py -i 25 -a 300 -o my_outdir
```


CHANGELOG.md
# combinator-FQ changelog

## 2023-06-16 edition

- Now, contig multiplicity is calculated simply as *cov_i* / *cov_1*.
- Now, LQ-coefficient is calculated with no regard to multiplicity of contigs.

### Version change:

`1.5.b -> 1.6.a`

## 2022-06-17 edition

- Added `.fna`, `.fsa`, and `.fsa_nt` to possible fasta extentions.

### Version change:

`1.5.a -> 1.5.b`

## 2021-05-12 edition

- combinator-FQ now can calculate expected genome length and multiplicity of contigs based on number of overlaps, if no coverage information is present in sequence headers.
- Minor output formatting fixes.

### Version changes:

`1.4.a -> 1.5.a`

## 2021-03-26 edition

- performed total refactoring. Memory usage is reduced. Made some cosmetic fixes.
- Moved combinator-FQ to the separate (this) repository from [https://github.com/masikol/cager-misc](https://github.com/masikol/cager-misc).

### Version changes:

`1.3.f -> 1.4.a`

## 2021-01-21 edition

- bug fix (combinator used to ternimate in the end if it couldn't extract ordinal number of contig from contig's name);

### Version changes:

`1.3.e -> 1.3.f`

## 2020-12-03 edition

- `-o` option addded.

### Version changes:

`1.3.d -> 1.3.e`

## 2020-10-08 edition

- combinator-FQ now rounds coverage with 2 trailing digits.

### Version changes:

`1.3.c --> 1.3.d`

## 2020-06-15 edition.

- errorneous handling of SPADEs's `NODE_1` with zero coverage fixed;

### Version changes:

`1.3.b --> 1.3.c`

## 2020-04-24 edition.

- compatibility with a5 improved;

### Version changes:

`1.3.a --> 1.3.b`

## 2020-04-15 edition.

- calculating of expected genome length is improved;

### Version changes:

`1.2.b --> 1.3.a`;

## 2020-04-10 evening edition.

- a5-compatibility bug fixed;

### Version changes:

`1.2.a --> 1.2.b`

## 2020-04-10 edition.

- calculating of expected genome length is embedded once again -- not considering contigs multiplicity;
- fasta-GC-content is added;

### Version changes:

`1.1.c --> 1.2.a`

## 2020-03-18 edition.

- calculating of expected genome length is disabled because it's impossible to handle high-copy replicons properly;

### Version changes:

`1.1.b --> 1.1.c`

## 2020-03-09 edition.

- calculation of expected genome length fixed;

### Version changes:

`1.1.a --> 1.1.b`

## 2020-03-07 edition.

- output file separated into three: `adjacent_contigs`. `full_matching_log` and `summary`;
- A5-coverage summary bug fixed;

### Version changes:

`1.0.a --> 1.1.a`

## 2020-03-06 edition.

- combinator-FQ version `1.0.a` added;


src/parse_args.py
# -*- encoding: utf-8 -*-

import os
import sys
import glob
import getopt
from typing import List, Sequence, Dict, Mapping, Any, Tuple

import src.filesystem
from src.print_help import print_help
from src.platform import platf_depend_exit


def parse_args(version: str, last_update_date: str) -> Tuple[Sequence[str], Mapping[str, Any]]:
    # Function parses command line arguments.
    # Returns two values:
    #  1. Collection of paths to input files.
    #  2. Dictionary of parameters (see function _parse_options).

    # Print help message and exit if required
    if '-h' in sys.argv[1:] or '--help' in sys.argv[1:]:
        print_help(version, last_update_date)
        platf_depend_exit()
    # end if

    # Print version and exit if required
    if '-v' in sys.argv[1:] or '--version' in sys.argv[1:]:
        print(version)
        platf_depend_exit()
    # end if

    # Parse arguments woth getopt
    opts: List[List[str]]
    args: List[str]
    try:
        opts, args = getopt.gnu_getopt(sys.argv[1:], 'hvk:i:a:o:',
            ['help', 'version', 'k-mer=', 'mink=', 'maxk=', 'outdir='])
    except getopt.GetoptError as err:
        print(str(err))
        platf_depend_exit(2)
    # end try

    # Extract paths to input files from parsed arguments
    contigs_fpaths: Sequence[str] = _get_input_fpaths(args)
    # Extract optional parameters from parsed arguments
    params: Dict[str, Any] = _parse_options(opts)

    # Verify mink and maxk:
    if params['i'] > params['a']:
        if '-i' not in sys.argv[1:] and '--mink' not in sys.argv[1:]:
            params['i'] = params['a']
        elif '-a' not in sys.argv[1:] and '--maxk' not in sys.argv[1:]:
            params['a'] = params['i']
        else:
            print('Error: minimum length of a k-mer is greater than maximum length of a k-mer.')
            print('Values specified by you:')
            print('Minimum length of a k-mer: {}.'.format(params['i']))
            print('Maximum length of a k-mer: {}.'.format(params['a']))
            platf_depend_exit(1)
        # end if
    # end if

    return contigs_fpaths, params
# end def parse_args


def _get_input_fpaths(args: Sequence[str]) -> Sequence[str]:
    # Function extracts paths to input files from `args` colection returned by `getopt.gnu_getopt`.
    # Returns collection of paths to input fasta files.

    contigs_fpaths: Sequence[str]

    if len(args) == 0:
        # If no input files are specified,
        #   search for input files in the working directory:

        # Get all fasta files in the workgin directory
        contigs_fpaths = tuple(
            filter(
                src.filesystem.is_fasta,
                glob.iglob(os.path.join(os.getcwd(), '*'))
            )
        )

        if len(contigs_fpaths) != 0:
            # Files are found -- ask for permission and proceed
            print('Following fasta files are found and will be processed:')
            i: int
            path: str
            for i, path in enumerate(contigs_fpaths):
                print(' {}. `{}`'.format(i+1, path))
            # end for
            error: bool = True
            while error:
                reply: str = input('Is it ok? Proceed? [y, n]:')
                if reply.lower() == 'y':
                    error = False
                elif reply.lower() == 'n':
                    print('Ok. Type `{} -h` for help.'.format(sys.argv[0]))
                    sys.exit(0)
                else:
                    print('Invalid reply: `{}`'.format(reply))
                # end if
            # end while
        else:
            # If there are no fasta files in the working directory -- exit
            print('Nothing to process.')
            print('Please, type `{} -h` for help.'.format(sys.argv[0]))
            platf_depend_exit(1)
        # end if
    else:
        # Check existance of input files
        arg: str
        for arg in args:
            if not os.path.exists(arg):
                print('Error: file `{}` does not exist.'.format(arg))
                platf_depend_exit(1)
            # end if
            if not src.filesystem.is_fasta(arg):
                print('Error: file `{}` is not a fasta file, considering it\'s extention.\n'
                      '  It if is indeed a fasta file, sorry for the inconvenience'
                      ' and please change it\'s extention to'
                      ' `.fa`, `.fasta`, `.fna`, `.fsa` or `.fsa_nt` (the file may be gzipped).'\
                    .format(arg))
                platf_depend_exit(1)
            # end if
        # end for
        contigs_fpaths = tuple(args)
    # end if

    # make all paths absolute
    contigs_fpaths = tuple(
        map(
            os.path.abspath,
            contigs_fpaths
        )
    )

    return contigs_fpaths
# end def _get_input_fpaths


def _parse_options(opts: Sequence[Sequence[str]]) -> Mapping[str, Any]:
    # Funtion extracts options from `args` colection returned by `getopt.gnu_getopt`.
    # Returns following dictionary:
    #    {
    #       'o': <outdir_path>,
    #       'i': <mink>,
    #       'a': <maxk>,
    #    }

    # Set default values for parameters
    params: Dict[str, Any] = {
        'o': os.path.join(os.getcwd(), 'combinator-result'), # outdir
        'i': 21,                                             # mink
        'a': 127,                                            # maxk
    }

    # Parse command line options
    opt: str
    arg: str
    for opt, arg in opts:

        # k-mer size
        if opt in ('-k', '--k-mer-len'):
            try:
                k: int = int(arg)
                if k <= 0:
                    raise ValueError
                # end if
            except ValueError:
                print('Error: length of a k-mer must be positive integer number.')
                print('Your value: `{}`'.format(arg))
                platf_depend_exit(1)
            else:
                params['i'], params['a'] = k, k
            # end try

        # Outdir
        elif opt in ('-o', '--outdir'):
            params['o'] = os.path.abspath(arg)

        # Minimim k-mer size
        elif opt in ('-i', '--mink'):
            try:
                params['i'] = int(arg)
                if params['i'] <= 0:
                    raise ValueError
                # end if
            except ValueError:
                print('Minimum length of a k-mer must be positive integer number.')
                print('Your value: `{}`'.format(arg))
                platf_depend_exit(1)
            # end try

        # Maximim k-mer size
        elif opt in ('-a', '--maxk'):
            try:
                params['a'] = int(arg)
                if params['a'] <= 0:
                    raise ValueError
                # end if
            except ValueError:
                print('Maximum length of k-mer must be positive integer number.')
                print('Your value: `{}`'.format(arg))
                platf_depend_exit(1)
            # end try
        # end if
    # end for

    return params
# end def _parse_options
